### PR TITLE
add sync-now action

### PR DESCRIPTION
**Pull request by sed-i**
This PR is for adding the `sync-now` action.
It is handy if an immediate update is needed from the repo into relation data.

Without this action, to force a sync, one would need to:
1. `juju config` to shorten the `git_wait` period
2. `juju model-config` to shorten the update-status interval
3. wait for relation data to update
4. restore the above to original values

Immediate use cases:
- integration tests: files show up on disk after the last hook fired - can use the action instead of the update-status hack
- similar for load tests: want to be able to run selenium/splinter knowing that the dashboards are already in place

### CODE REVIEW

**Review by rbarry82 — APPROVED** [review]

File: src/charm.py [hunk 1]
```diff
@@ -303,19 +247,71 @@ def _remove_repo_folder(self):
         # but to keep unittest simpler, doing it from the charm container's mount point
         shutil.rmtree(self._repo_path, ignore_errors=True)
 
+    def _command(self, one_time=False) -> List[str]:
```
Comment (Abuelodelanada) [inline src/charm.py:250]: Isn't this method name to common?

**Review by Abuelodelanada — APPROVED** [review]

File: src/charm.py (hunk 1 above)
Comment (simskij) [reply src/charm.py:250]: I agree

File: src/charm.py [hunk 2]
```diff
@@ -291,6 +201,40 @@ def _common_exit_hook(self) -> None:  # noqa: C901
...
+        try:
+            stdout, warnings = process.wait_output()
+        except ExecError as e:
+            for line in e.stderr.splitlines():
+                event.log(line)
+            event.fail("Exited with code {e.exit_code}.")
+            return
+        except ChangeError as e:
+            event.fail(str(e))
+            return
```
Comment (simskij) [inline src/charm.py:227]: Extract try/catch pairs into separate (private) methods.

File: src/charm.py [hunk 3]
```diff
@@ -303,19 +247,71 @@ def _remove_repo_folder(self):
...
+        cmd = ["/git-sync"]
+        cmd.extend(["--repo", repo])
+        if branch:
+            cmd.extend(["--branch", branch])
+        if rev:
+            cmd.extend(["--rev", rev])
+        cmd.extend(
+            [
+                "--depth",
+                "1",
```
Comment (simskij) [inline src/charm.py:271]: why a depth of one? this requires explaining

**Review by simskij — DISMISSED** [review]

File: src/charm.py (hunk 3 above)
Comment (sed-i) [reply src/charm.py:271]: There is already an issue for that :)
https://github.com/canonical/cos-configuration-k8s-operator/issues/7

File: src/charm.py (hunk 1 above)
Comment (sed-i) [reply src/charm.py:250]: What do you mean?

File: src/charm.py (hunk 2 above)
Comment (sed-i) [reply src/charm.py:227]: What do you mean?

File: src/charm.py [hunk 4]
```diff
@@ -303,19 +247,71 @@ def _remove_repo_folder(self):
...
+        return Layer(
+            {
+                "summary": f"{self._service_name} layer",
+                "description": f"pebble config layer for {self._service_name}",
+                "services": {
+                    self._service_name: {
+                        "override": "replace",
+                        "summary": f"{self._service_name} service",
+                        "startup": "disabled",
+                        "command": " ".join(self._command()),
```
[the remark this answers is not in the document]
Comment (sed-i) [reply src/charm.py:310]: It used to, until I also needed it as `List[str]` for `container.exec`.

File: src/charm.py (hunk 1 above)
Comment (sed-i) [reply src/charm.py:250]: Fixed in next commit.

**Review by rbarry82 — APPROVED** [review]

**Pull request merged**